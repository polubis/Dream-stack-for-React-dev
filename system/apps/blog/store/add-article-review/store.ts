import { create } from 'zustand';
import type { AddArticleReviewStore } from './defs';
import { addArticleReviewForm } from './core';

const useAddArticleReviewStore = create<AddArticleReviewStore.State>(() => ({
  is: 'idle',
  form: addArticleReviewForm.init({ content: '' }),
}));

export { useAddArticleReviewStore, addArticleReviewForm };
