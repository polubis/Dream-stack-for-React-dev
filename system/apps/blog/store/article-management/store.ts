import { create } from 'zustand';
import type * as ArticleManagement from './defs';

const useArticleManagementStore = create<ArticleManagement.State>(() => ({
  is: 'idle',
}));

export { useArticleManagementStore };
