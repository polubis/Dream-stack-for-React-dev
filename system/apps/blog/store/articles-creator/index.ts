export * from './articles-creator.store';
export * from './defs';
