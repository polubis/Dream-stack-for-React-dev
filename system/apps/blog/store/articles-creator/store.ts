import { create } from 'zustand';
import { ArticlesCreator } from './defs';
import { idle } from './states';

const useArticlesCreatorStore = create<ArticlesCreator.State>(idle);

export { useArticlesCreatorStore };
