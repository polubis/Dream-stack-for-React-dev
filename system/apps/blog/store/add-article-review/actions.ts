import { createArticleReview, getError } from '@system/blog-api';
import type * as AddArticleReview from './defs';
import { addArticleReviewForm, useAddArticleReviewStore } from './store';

const { setState: set, getState: get } = useAddArticleReviewStore;

const add_article_review_actions: AddArticleReview.Actions = {
  setField: (key, value) => {
    set({ form: addArticleReviewForm.set(get().form)({ [key]: value }) });
  },
  confirm: async (id) => {
    const form = addArticleReviewForm.confirm(get().form);

    if (form.invalid) {
      set({ form });
      return;
    }

    set({ is: 'busy' });

    try {
      const response = await createArticleReview({
        id,
        ...get().form.values,
      });

      set({ is: 'ok' });
      add_article_review_actions.setField('content', '');

      return response;
    } catch (error: unknown) {
      set({ is: 'fail', error: getError(error) });
    }
  },
};

export { add_article_review_actions };
