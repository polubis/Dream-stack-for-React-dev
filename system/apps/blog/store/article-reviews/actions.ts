import { getArticleReviews, getError } from '@system/blog-api';
import type * as ArticleReviews from './defs';
import { useArticleReviewsStore } from './store';

const { setState: set, getState: get } = useArticleReviewsStore;

const article_reviews_actions: ArticleReviews.Actions = {
  load: async (id) => {
    set({ is: 'busy' }, true);

    try {
      const { data } = await getArticleReviews({ id });
      set({ is: 'ok', reviews: data }, true);
    } catch (error: unknown) {
      set({ is: 'fail', error: getError(error) }, true);
    }
  },
  addReview: (review) => {
    const state = get();

    if (state.is === 'ok') {
      set({ reviews: [...state.reviews, review] });
      return;
    }

    throw Error('You are trying to add review when there is no data yet');
  },
  reset: () => {
    set({ is: 'idle' }, true);
  },
};

export { article_reviews_actions };
