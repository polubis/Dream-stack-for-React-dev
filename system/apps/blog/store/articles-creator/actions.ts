import type { Lang, Url } from '@system/blog-api-models';
import { useArticlesCreatorStore } from './store';
import {
  createFail,
  created,
  creating,
  creation,
  editFail,
  edited,
  editing,
  edition,
  idle,
  loadFail,
  loading,
} from './states';
import {
  createArticle,
  getArticle,
  getError,
  updateArticle,
} from '@system/blog-api';
import { ArticlesCreator } from './defs';
import { creatorForm, initForm } from './form';

const validateConfirm = (
  state: ArticlesCreator.Creation | ArticlesCreator.Edition
): void | never => {
  const { form } = state;
  const { thumbnail } = form.values;

  if (form.invalid) throw Error('You tried to send invalid form data');

  if (!thumbnail.file) throw Error('File is required');
};

const store = () => ({
  setState: (state: ArticlesCreator.State) =>
    useArticlesCreatorStore.setState(state, true),
  getState: useArticlesCreatorStore.getState,
});

const { setState, getState } = store();

const articles_creator_actions = {
  reset: (): void => {
    setState(idle());
  },
  init: async (lang: Lang, url?: Url) => {
    if (url) {
      setState(loading(url));

      try {
        const { data: article } = await getArticle({ url, lang });
        const { title, description, content, thumbnailUrl } = article;

        setState(
          edition(
            url,
            initForm(lang, {
              title,
              description,
              thumbnail: {
                file: null,
                preview: thumbnailUrl,
              },
              content,
            }),
            article
          )
        );
      } catch (error: unknown) {
        setState(loadFail(getError(error)));
      }

      return;
    }

    setState(creation(initForm(lang)));
  },
  change: <
    K extends keyof ArticlesCreator.FormData,
    V extends ArticlesCreator.FormData[K]
  >(
    key: K,
    value: V
  ) => {
    const s = getState();

    if (s.is === 'creation') {
      setState(creation(creatorForm.set(s.form)({ [key]: value })));
      return;
    }

    if (s.is === 'edition') {
      setState(
        edition(s.url, creatorForm.set(s.form)({ [key]: value }), s.article)
      );
      return;
    }

    setState(
      loadFail(getError('The current state is not changeable: ' + s.is))
    );
  },
  confirm: async () => {
    const s = getState();

    if (s.is === 'creation') {
      const { form } = s;
      const { title, description, content, thumbnail, lang } = form.values;

      try {
        validateConfirm(s);

        setState(creating(form));

        await createArticle({
          title,
          description,
          content,
          lang,
          thumbnail: thumbnail.file,
        });

        setState(created(form));
      } catch (error: unknown) {
        setState(createFail(s.form, getError(error)));
      }
      return;
    }

    if (s.is === 'edition') {
      const { form, url, article } = s;
      const { title, description, content, thumbnail, lang } = form.values;

      try {
        validateConfirm(s);

        setState(editing(url, form, article));

        await updateArticle({
          url,
          title,
          description,
          content,
          lang,
          thumbnail: thumbnail.file,
        });

        setState(edited(url, form, article));
      } catch (error: unknown) {
        setState(editFail(url, form, getError(error), article));
      }
      return;
    }

    setState(
      loadFail(getError('The current state is not confirmable: ' + s.is))
    );
  },
};

export { articles_creator_actions };
