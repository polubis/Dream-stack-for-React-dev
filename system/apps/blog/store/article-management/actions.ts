import { useArticleManagementStore } from './store';
import type * as ArticleManagement from './defs';
import { article_reviews_actions } from '../article-reviews';
import { article_store_actions } from '../article/actions';
import { add_article_review_store_actions } from '../add-article-review';
import { mockArticleReview } from '@system/blog-api-mocks';
import { change_article_status_actions } from '../change-article-status';

const { setState: set } = useArticleManagementStore;

const article_management_actions: ArticleManagement.Actions = {
  start: (id, url, lang) => {
    set({ is: 'active', id });
    article_reviews_actions.load(id);
    article_store_actions.load({ url, lang });
  },
  sendForApproval: async (id) => {
    await change_article_status_actions.sendForApproval(id);
    article_store_actions.update({ status: 'WaitingForApproval' });
    return;
  },
  reset: () => {
    set({ is: 'idle' });
  },
  confirm: async (id) => {
    await add_article_review_store_actions.confirm(id);
    // @TODO: Remove mock when backend will finish work.
    article_reviews_actions.addReview(mockArticleReview());
  },
  changeStatus: async (id, status) => {
    if (status === 'Accepted') {
      await change_article_status_actions.accept(id);
      article_store_actions.update({ status });
      return;
    }

    if (status === 'NeedWork') {
      await change_article_status_actions.reject(id);
      article_store_actions.update({ status });
      return;
    }
  },
};

export { article_management_actions };
