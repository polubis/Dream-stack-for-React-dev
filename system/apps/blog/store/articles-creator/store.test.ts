import { act } from '@testing-library/react';
import { useArticlesCreatorStore } from './store';
import { storeFixture } from '../test-utils';
import { articles_creator_actions } from './actions';

describe('Articles creation feature works when: ', () => {
  it('initial state is valid', () => {
    const { result, restore } = storeFixture(useArticlesCreatorStore);

    expect(result.current).toMatchSnapshot();

    restore();
  });

  it('initializes the form', () => {
    const { result, restore } = storeFixture(useArticlesCreatorStore);

    act(() => {
      articles_creator_actions.init('en');
    });

    expect(result.current).toMatchSnapshot();

    restore();
  });

  it('throws an exception when someone tries to change form field in not allowed state', () => {
    const { result, restore } = storeFixture(useArticlesCreatorStore);

    act(() => {
      articles_creator_actions.change('description', 'some-text');
    });

    expect(result.current).toMatchSnapshot();
    restore();
  });

  describe('throws an error during confirmation', () => {
    it('if thumbnail is missing', () => {
      const { result, restore } = storeFixture(useArticlesCreatorStore);

      act(() => {
        articles_creator_actions.init('en');
        articles_creator_actions.confirm();
      });

      expect(result.current).toMatchSnapshot();

      restore();
    });

    it('if form is confirmed when invalid', () => {
      const { result, restore } = storeFixture(useArticlesCreatorStore);

      act(() => {
        articles_creator_actions.init('en');
        articles_creator_actions.change('title', '');
        articles_creator_actions.confirm();
      });

      expect(result.current).toMatchSnapshot();

      restore();
    });

    it('if form is confirmed in not allowed state', () => {
      const { result, restore } = storeFixture(useArticlesCreatorStore);

      act(() => {
        articles_creator_actions.confirm();
      });

      expect(result.current.is).toBe('load-fail');
      expect(result.current).toMatchSnapshot();

      restore();
    });
  });

  it('allows to change and validate form fields', () => {
    const { result, restore } = storeFixture(useArticlesCreatorStore);

    act(() => {
      articles_creator_actions.init('en');
    });

    expect(result.current.is).toBe('creation');

    act(() => {
      articles_creator_actions.change('description', 'some-text');
    });

    expect(result.current).toMatchSnapshot();

    restore();
  });
});
