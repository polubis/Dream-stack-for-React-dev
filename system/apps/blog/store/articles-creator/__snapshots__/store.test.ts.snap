// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles creation feature works when:  allows to change and validate form fields 1`] = `
Object {
  "form": Object {
    "dirty": false,
    "errors": Object {
      "content": null,
      "description": null,
      "lang": null,
      "thumbnail": null,
      "title": "required",
    },
    "invalid": true,
    "invalidCount": 1,
    "keys": Array [
      "title",
      "description",
      "thumbnail",
      "content",
      "lang",
    ],
    "pristine": true,
    "progress": 80,
    "result": Object {
      "content": false,
      "description": false,
      "lang": false,
      "thumbnail": false,
      "title": true,
    },
    "touch": Object {
      "content": false,
      "description": true,
      "lang": false,
      "thumbnail": false,
      "title": false,
    },
    "touched": true,
    "untouched": false,
    "valid": false,
    "validCount": 4,
    "values": Object {
      "content": "",
      "description": "some-text",
      "lang": "en",
      "thumbnail": Object {
        "file": null,
        "preview": "",
      },
      "title": "",
    },
  },
  "is": "creation",
}
`;

exports[`Articles creation feature works when:  initial state is valid 1`] = `
Object {
  "is": "idle",
}
`;

exports[`Articles creation feature works when:  initializes the form 1`] = `
Object {
  "form": Object {
    "dirty": false,
    "errors": Object {
      "content": null,
      "description": null,
      "lang": null,
      "thumbnail": null,
      "title": null,
    },
    "invalid": false,
    "invalidCount": 0,
    "keys": Array [
      "title",
      "description",
      "thumbnail",
      "content",
      "lang",
    ],
    "pristine": true,
    "progress": 0,
    "result": Object {
      "content": false,
      "description": false,
      "lang": false,
      "thumbnail": false,
      "title": false,
    },
    "touch": Object {
      "content": false,
      "description": false,
      "lang": false,
      "thumbnail": false,
      "title": false,
    },
    "touched": false,
    "untouched": true,
    "valid": true,
    "validCount": 5,
    "values": Object {
      "content": "",
      "description": "",
      "lang": "en",
      "thumbnail": Object {
        "file": null,
        "preview": "",
      },
      "title": "",
    },
  },
  "is": "creation",
}
`;

exports[`Articles creation feature works when:  throws an error during confirmation if form is confirmed in not allowed state 1`] = `
Object {
  "error": Object {
    "key": "unknown",
    "message": "Something went wrong...",
  },
  "is": "load-fail",
}
`;

exports[`Articles creation feature works when:  throws an error during confirmation if form is confirmed when invalid 1`] = `
Object {
  "error": Object {
    "key": "unknown",
    "message": "Something went wrong...",
  },
  "form": Object {
    "dirty": false,
    "errors": Object {
      "content": null,
      "description": null,
      "lang": null,
      "thumbnail": null,
      "title": "required",
    },
    "invalid": true,
    "invalidCount": 1,
    "keys": Array [
      "title",
      "description",
      "thumbnail",
      "content",
      "lang",
    ],
    "pristine": true,
    "progress": 80,
    "result": Object {
      "content": false,
      "description": false,
      "lang": false,
      "thumbnail": false,
      "title": true,
    },
    "touch": Object {
      "content": false,
      "description": false,
      "lang": false,
      "thumbnail": false,
      "title": true,
    },
    "touched": true,
    "untouched": false,
    "valid": false,
    "validCount": 4,
    "values": Object {
      "content": "",
      "description": "",
      "lang": "en",
      "thumbnail": Object {
        "file": null,
        "preview": "",
      },
      "title": "",
    },
  },
  "is": "create-fail",
}
`;

exports[`Articles creation feature works when:  throws an error during confirmation if thumbnail is missing 1`] = `
Object {
  "error": Object {
    "key": "unknown",
    "message": "Something went wrong...",
  },
  "form": Object {
    "dirty": false,
    "errors": Object {
      "content": null,
      "description": null,
      "lang": null,
      "thumbnail": null,
      "title": null,
    },
    "invalid": false,
    "invalidCount": 0,
    "keys": Array [
      "title",
      "description",
      "thumbnail",
      "content",
      "lang",
    ],
    "pristine": true,
    "progress": 0,
    "result": Object {
      "content": false,
      "description": false,
      "lang": false,
      "thumbnail": false,
      "title": false,
    },
    "touch": Object {
      "content": false,
      "description": false,
      "lang": false,
      "thumbnail": false,
      "title": false,
    },
    "touched": false,
    "untouched": true,
    "valid": true,
    "validCount": 5,
    "values": Object {
      "content": "",
      "description": "",
      "lang": "en",
      "thumbnail": Object {
        "file": null,
        "preview": "",
      },
      "title": "",
    },
  },
  "is": "create-fail",
}
`;

exports[`Articles creation feature works when:  throws an exception when someone tries to change form field in not allowed state 1`] = `
Object {
  "error": Object {
    "key": "unknown",
    "message": "Something went wrong...",
  },
  "is": "load-fail",
}
`;
